DESIGN.md

```diff
@@ -0,0 +1,412 @@
+# COSE Golang Library
+
+## Objectives
+
+The objectives of the `go-cose` library of this doc are
+
+- Implement `COSE_Sign` and `COSE_Sign1` as specified in [RFC8152](https://datatracker.ietf.org/doc/html/rfc8152).
+- Make the implementation secure and efficient.
+- Provide signing and verification based on golang built-in crypto libraries.
+- Provide extensibility so that other implementation can be used for signing and verification.
+  - This objective implies remote signing.
+- Be golang native.
+  - Write documentation and examples for `godoc`.
+  - Test using `go test` with both positive and negative test cases.
+
+## Implementation Gaps
+
+Here are the gaps between the current implementation and the objectives:
+
+- The current implementation implements `COSE_Sign` and `COSE_Sign1`. However, the verification process does not follow [RFC8152](https://datatracker.ietf.org/doc/html/rfc8152) and causes [issue #7](https://github.com/veraison/go-cose/issues/7). There are also implementation errors in signing and verification (see [issue #8](https://github.com/veraison/go-cose/issues/8)).
+- The efficiency of the current implementation can be improved with crypto domain knowledge. For instances,
+  - Remove the use of constant time copy in `I2OSP`.
+  - Remove the check on `(r,s)` against `n`.
+- The current implementation implements `PS256`, `ES256`, `ES384`, `ES512`, and misses other common algorithms like `PS384`, `PS512`, `EdDSA` (e.g. `Ed25519`).
+  - All conventional signature schemes are _signature with appendix_.
+  - `RS256`, `RS384`, `RS512`, `ES256K` are marked **NOT RECOMMENDED** by [RFC8812](https://datatracker.ietf.org/doc/html/rfc8812).
+- The current implementation has solid `Signer` and `Verify` structures and can only sign or verify against certain algorithms (see [issue #9](https://github.com/veraison/go-cose/issues/9)). The implementation cannot be extended to have new algorithms or new implementation for existing algorithms.
+- The current implementation is not golang native.

```

Comment:
Is Line 28 referring to veraison/go-cose or some other repo? when it refers to golang nativity ? 

As far as I reckon that was with older notary implementation ? If so needs to be specified clearly.

Reply:
@shizhMSFT, what do you think?

Reply:
updated in 6ef3dd4